crates/turbopack-build/src/ecmascript/node/evaluate/runtime.rs

```diff
@@ -0,0 +1,178 @@
+use turbopack_core::{
+    asset::{Asset, AssetContentVc, AssetVc},
+    chunk::ChunkingContext,
+    code_builder::{CodeBuilder, CodeVc},
+    ident::{AssetIdent, AssetIdentVc},
+    reference::{AssetReference, AssetReferenceVc, AssetReferencesVc},
+    resolve::{ResolveResult, ResolveResultVc},
+    source_map::{
+        GenerateSourceMap, GenerateSourceMapVc, OptionSourceMapVc, SourceMapAssetReferenceVc,
+    },
+    virtual_fs::VirtualFileSystemVc,
+};
+use turbopack_ecmascript::utils::StringifyJs;
+
+use crate::BuildChunkingContextVc;
+
+/// An Ecmascript chunk that contains the Node.js runtime code.
+#[turbo_tasks::value(shared)]
+pub(crate) struct EcmascriptBuildNodeRuntimeChunk {
+    chunking_context: BuildChunkingContextVc,
+}
+
+#[turbo_tasks::value_impl]
+impl EcmascriptBuildNodeRuntimeChunkVc {
+    /// Creates a new [`EcmascriptBuildNodeRuntimeChunkVc`].
+    #[turbo_tasks::function]
+    pub fn new(chunking_context: BuildChunkingContextVc) -> Self {
+        EcmascriptBuildNodeRuntimeChunk { chunking_context }.cell()
+    }
+
+    #[turbo_tasks::function]
+    async fn code(self) -> Result<CodeVc> {
+        let this = self.await?;
+
+        let output_root = this.chunking_context.output_root().await?;
+        let runtime_path = self.ident().path().await?;
+        let runtime_public_path = if let Some(path) = output_root.get_path_to(&runtime_path) {
+            path
+        } else {
+            bail!(
+                "runtime path {} is not in output root {}",
+                runtime_path.to_string(),
+                output_root.to_string()
+            );
+        };
+
+        let mut code = CodeBuilder::default();
+
+        writedoc!(
+            code,
+            r#"
+                const RUNTIME_PUBLIC_PATH = {};
+            "#,
+            StringifyJs(&*runtime_public_path)
+        )?;
+
+        // TODO(alexkirsz) Process asset.
+        let runtime_code: turbo_tasks::ReadRef<FileContent> =
+            turbopack_ecmascript_runtime::embed_file("build/runtime.ts").await?;
+
+        match &*runtime_code {
+            FileContent::NotFound => bail!("shared runtime code is not found"),
+            FileContent::Content(file) => code.push_source(file.content(), None),
+        };
+
+        Ok(CodeVc::cell(code.build()))
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl ValueToString for EcmascriptBuildNodeRuntimeChunk {
+    #[turbo_tasks::function]
+    async fn to_string(&self) -> Result<StringVc> {
+        Ok(StringVc::cell(
+            "Ecmascript Build Node Runtime Chunk".to_string(),
+        ))
+    }
+}
+
+#[turbo_tasks::function]
+fn runtime_fs() -> FileSystemVc {
+    VirtualFileSystemVc::new().into()
+}
+
+#[turbo_tasks::value_impl]
+impl Asset for EcmascriptBuildNodeRuntimeChunk {
+    #[turbo_tasks::function]
+    fn ident(&self) -> AssetIdentVc {
+        let ident = AssetIdentVc::new(Value::new(AssetIdent {
+            path: runtime_fs().root().join("runtime.js"),
```

Comment:
Not a huge fan of having an extra filesystem for that. Could we just put that into the embedded filesystem. No worries if the file doesn't exist there. We already do that a bunch of times in other places.

It's more clear when every turbopack file is inside of `[turbopack]/`